### ISSUE TITLE
`entry.chunkLoading = false` is overridden by `SplitChunksPlugin`

### ISSUE THREAD

**Issue by srmagura** [webpack/webpack#16320]
# Bug report

Closely related: https://github.com/webpack/webpack/issues/4819.

**What is the current behavior?**

Here is an example Webpack config, with two entry points `app` and `serviceWorker`. For `app`, the config uses `SplitChunksPlugin` to put `node_modules` in a "vendor" chunk to improve long-term caching. Trying to dynamically load chunks from a service worker has failed each time I have tested it, so I want the `serviceWorker` entry to be compiled to one big JS file.

```js
const path = require("path");

module.exports = {
  mode: "development",
  entry: {
    app: path.resolve(__dirname, "src/app.js"),
    serviceWorker: {
      import: path.resolve(__dirname, "src/serviceWorker.js"),
      chunkLoading: false, // This _should_ put everything in one JS file, but it doesn't
    },
  },

  output: {
    clean: true,
  },
  devtool: false,

  // Enable long-term caching of node_modules. See https://webpack.js.org/guides/caching/
  optimization: {
    moduleIds: "deterministic",
    splitChunks: {
      cacheGroups: {
        vendor: {
          test: /[\\/]node_modules[\\/]/,
          name: "vendor",

          // Include all node_modules that are required for the initial load
          // in one big chunk. Dynamically-imported node_modules will still be
          // put in their own chunks.
          chunks: "initial",
        },
      },
    },
  },
};
```

The Webpack docs state [here](https://webpack.js.org/configuration/entry-context/#entry-descriptor) that you can pass `chunkLoading: false` when defining an entry to put everything in the main chunk. It seems like `SplitChunksPlugin` ignores this option.

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Clone my repo: https://github.com/srmagura/webpack-split-chunks-repro
2. `yarn install`
3. `yarn build`
4. Check `dist/serviceWorker.js` and observe that `lodash` is not in the file. Lodash is loaded dynamically from the `vendor` chunk.

**What is the expected behavior?**

`chunkLoading: false` should cause `serviceWorker.js` to include all chunks required for the service worker to run. This means Lodash should be in `serviceWorker.js`.

**Other relevant information:**

As a workaround, I could compile my service worker through a separate Webpack config, but I wanted to check if there was a better option before doing so.

webpack version:  5.74.0
Node.js version: 16.17.0
Operating System: Ubuntu 22.04

**Comment by rahulkumar26** [comment]
This is interesting!

If I understand correctly,

- you have one npm module and you want to add this module to each entry. Then what's the reson for using SplitChunks? 
- Or, may be you want to just add lodash to each entries and other modules (in future) to be part of Vendor?

if the use case is the 2nd one then why not adding `test: /[\\/]node_modules[\\/]!(lodash)[\\/]/`

**Comment by srmagura** [comment]
Hey @rahulkumar26, in reality both entrypoints depend on many more npm modules than just `lodash`. I used only `lodash` in the example to keep things simple.

I want to have a separate vendors chunk for the `app` entry to take advantage of long-term caching. But the way Webpack loads additional chunks seems to be incompatible with service worker, which is why the service worker must be a single JS file.

**Comment by alexander-akait** [comment]
You can't do it, because it is a limitation, you can move service worker bundling to own compilation (i.e. multi compiler mode with `target: 'webworker'`), otherwise webpack just don't know how to load chunk

**Comment by srmagura** [comment]
Thanks, yes that was my backup plan. Could this limitation be removed, or should I close the issue?

**Comment by alexander-akait** [comment]
I will look in near future on code deeply and I will say, so let's keep open now

**Comment by alexander-akait** [comment]
@srmagura what about this:
```js
const path = require("path");

module.exports = {
  mode: "development",
  entry: {
    app: path.resolve(__dirname, "src/app.js"),
    serviceWorker: {
      layer: 'worker',
      import: path.resolve(__dirname, "src/serviceWorker.js"),
    },
  },

  output: {
    clean: true,
  },
  devtool: false,
  experiments: {
    layers: true
  },

  // Enable long-term caching of node_modules. See https://webpack.js.org/guides/caching/
  optimization: {
    moduleIds: "deterministic",
    splitChunks: {
      cacheGroups: {
        vendor: {
          layer: (name) => {
            name === "worker"
          },
          test: /[\\/]node_modules[\\/]/,
          name: "vendor",

          // Include all node_modules that are required for the initial load
          // in one big chunk. Dynamically-imported node_modules will still be
          // put in their own chunks.
          chunks: "initial",
        },
      },
    },
  },
};
```

**Comment by alexander-akait** [comment]
So you keep splitting from application (and async chunks) and disable code splitting for `worker` layer

**Comment by srmagura** [comment]
Awesome!!! That works. Thank you so much.

**Comment by srmagura** [comment]
For anyone who may be reading this — the layers thing _seemed_ to be working but actually broke parts of our app. Specifically, we were seeing React context values not being passed down correctly to code in dynamically-loaded chunks. My best guess is that the React JavaScript got included in multiple chunks.

We've switched to just using a separate Webpack config for the service worker, since this seems less error prone and easier to understand.

**Comment by alexander-akait** [comment]
@srmagura Can you provide example of the problem?

**Comment by srmagura** [comment]
Hey @alexander-akait, here is some information about the problem. I've attached our full Webpack config as well as a webpack-bundle-analyzer screenshot that shows that react-dom has been included in multiple chunks. By the way, I forgot to mention that the problem only occurs in production builds.

Let me know if you think it is worthwhile for me to create a minimal repro for this. 


<details>
   <summary>webpack.config.cjs</summary>

```js
const path = require('path');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin');
const ReactRefreshWebpackPlugin = require('@pmmmwh/react-refresh-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const DotenvPlugin = require('dotenv-webpack');
const getCSSModuleLocalIdent = require('react-dev-utils/getCSSModuleLocalIdent');
const { DefinePlugin } = require('webpack');
const CopyPlugin = require('copy-webpack-plugin');
const { BundleAnalyzerPlugin } = require('webpack-bundle-analyzer');

// Remember to change this back to false before merging
const RUN_SERVICE_WORKER_IN_DEV = false;

module.exports = (env, argv) => {
  const mode = argv.mode ?? 'development';
  const production = mode === 'production';

  const plugins = [
    new HtmlWebpackPlugin({
      template: path.resolve(__dirname, 'public/index.html'),
    }),

    new CopyPlugin({
      patterns: [
        {
          from: 'public',
          filter: path => !path.endsWith('index.html'),
        },
        {
          from: 'compiled-lang',
          to: 'compiled-lang',
          filter: path => path.endsWith('.json'),
        },
      ],
    }),

    new ForkTsCheckerWebpackPlugin({ typescript: { memoryLimit: 4096 } }),

    new MiniCssExtractPlugin({
      filename: '[name].[contenthash].css',
    }),

    new DotenvPlugin(
      production
        ? { path: path.resolve(__dirname, '.env.production') }
        : undefined,
    ),

    new DefinePlugin({
      'process.env.REACT_APP_VERSION': JSON.stringify(
        env.REACT_APP_VERSION ?? 'dev',
      ),
      'process.env.RUN_SERVICE_WORKER_IN_DEV': JSON.stringify(
        RUN_SERVICE_WORKER_IN_DEV,
      ),
    }),
  ];

  if (env.bundle_analyzer) {
    plugins.push(new BundleAnalyzerPlugin());
  }

  if (!production) {
    plugins.push(new ReactRefreshWebpackPlugin());
  }

  const entry = {
    app: { layer: 'app', import: path.resolve(__dirname, 'src/index.tsx') },
  };

  // serviceWorker is only compiled in production mode because having multiple
  // entries per page mucks with hot module reloading
  if (production || RUN_SERVICE_WORKER_IN_DEV) {
    entry.serviceWorker = {
      layer: 'serviceWorker',
      import: path.resolve(__dirname, 'src/serviceWorker.ts'),
    };
  }

  return {
    mode,
    entry,
    resolve: {
      extensions: ['.ts', '.tsx', '.js', '.json'],
      modules: [path.resolve(__dirname, '.'), 'node_modules'],

      alias: {
        // we use a custom version of recast.js built from this codebase:
        // https://github.com/spotxyz/Extensions/tree/master/recastjs
        'recast-detour': path.resolve(__dirname, './static/recast.js'),
      },

      fallback: {
        // recast-detour tries to import Node.js modules but we use it in the
        // browser
        path: false,
        fs: false,

        // For slugify
        crypto: false,
      },
    },
    module: {
      rules: [
        {
          test: /\.tsx?$/,

          // Consider switching from Babel back to swc if @formatjs/swc-plugin
          // is ever fixed
          // (see https://github.com/formatjs/formatjs/issues/3782)
          use: require.resolve('babel-loader'),
        },

        // https://github.com/graphql/graphql-js/issues/2721#issuecomment-723008284
        {
          test: /\.m?js$/,
          resolve: {
            fullySpecified: false,
          },
        },
        {
          test: /\.s?css$/,
          use: [
            {
              loader: !production
                ? 'style-loader'
                : MiniCssExtractPlugin.loader,
            },
            {
              loader: 'css-loader',
              options: {
                sourceMap: true,
                importLoaders: 2,
                modules: {
                  auto: /\.module\.scss/,
                  mode: 'local',
                  getLocalIdent: getCSSModuleLocalIdent,
                },
              },
            },
            {
              loader: 'postcss-loader',
              options: {
                postcssOptions: {
                  plugins: ['autoprefixer'],
                },
              },
            },
            {
              loader: 'sass-loader',
              options: {
                sourceMap: true,
              },
            },
          ],
        },
        {
          test: /\.svg$/,
          use: [
            {
              loader: require.resolve('@svgr/webpack'),
              options: {
                prettier: false,
                svgo: false,
                svgoConfig: {
                  plugins: [{ removeViewBox: false }],
                },
                titleProp: true,
                ref: true,
              },
            },
            {
              loader: require.resolve('file-loader'),
              options: {
                name: 'static/media/[name].[hash].[ext]',
              },
            },
          ],
          issuer: {
            and: [/\.(ts|tsx|js|jsx|md|mdx)$/],
          },
        },
        {
          test: /\.(frag|vert)$/,
          type: 'asset/source',
        },
        {
          test: /\.(png|jpg|mp3|wav)$/,
          type: 'asset/resource',
        },
      ],
    },
    output: {
      filename: ({ runtime }) => {
        // Service worker path must not contain a hash, otherwise multiple
        // service workers will be installed

        // For production: chunkhash is necessary to prevent the service worker
        // from returning a cached file when the code has changed.
        //
        // For development: you aren't supposed to use hashes in filenames
        // because it causes too many files to build up in Webpack's cache.
        // https://github.com/webpack/webpack/discussions/16222

        return production && runtime !== 'serviceWorker'
          ? '[name].[chunkhash].js'
          : '[name].js';
      },

      chunkFilename: production ? '[id].[chunkhash].js' : '[id].js',

      path: path.resolve(__dirname, 'dist'),
      publicPath: '/',
      clean: true,
      devtoolModuleFilenameTemplate: 'webpack:///[absolute-resource-path]',
    },

    experiments: {
      layers: true,
    },

    // Enable long-term caching of node_modules. See https://webpack.js.org/guides/caching/
    optimization: {
      moduleIds: 'deterministic',
      splitChunks: {
        cacheGroups: {
          vendor: {
            // DO NOT CODESPLIT THE SERVICE WORKER. It will error when the
            // entrypoint tries to load the additional chunks.
            //
            // Sometimes `layer` is null or undefined. I'm not sure why this is,
            // since Webpack has virtually no documentation on this feature.
            // Without the `!layer` check, @babylonjs/core will be included in
            // both app.js and vendor.js which will cause issues.
            layer: layer => !layer || layer === 'app',

            test: /[\\/]node_modules[\\/]/,
            name: 'vendor',

            // Include all node_modules that are required for the initial load
            // in one big chunk. Dynamically-imported node_modules will still be
            // put in their own chunks.
            chunks: 'initial',
          },
        },
      },
    },

    devtool: production ? 'source-map' : 'cheap-module-source-map',
    plugins,
    devServer: {
      port: 3002,
      allowedHosts: 'all',
      historyApiFallback: true,
      client: {
        // We can't have Caddy and webpack-dev-server listening on the same
        // port, so Caddy listens on 3005 and webpack-dev-server listens on
        // 3002.
        webSocketURL: 'wss://spotd.xyz:3005/ws',
      },

      // Service worker + hot module reloading = UTTER CHAOS
      webSocketServer: RUN_SERVICE_WORKER_IN_DEV ? false : undefined,
    },
  };
};
```
</details>

![webpack-layers-problem](https://user-images.githubusercontent.com/801549/196721649-3cd05238-0845-484c-a573-95a6015183cb.png)

**Comment by alexander-akait** [comment]
@srmagura Sound like you have a module which used react-dom in the service worker and in the app, is it right?

**Comment by srmagura** [comment]
No, react-dom is never imported by the service worker. The app is split into multiple chunks, and there are multiple chunks that import react-dom.

**Comment by alexander-akait** [comment]
@srmagura Is it possible that another module which used in service worker uses `react-dom`? Can you define https://github.com/webpack/webpack/blob/main/lib/config/defaults.js#L1196 and set `layer` for this too? Also set `default: false` and `defaultVendors: false`, so we disable the default split chunks in favor our logic with layer app (if you need vendor for service worker or dynamic chunks you can set another app)

**Comment by srmagura** [comment]
Oh, and I'm positive that the service worker does not depend on react-dom whatsoever. It imports a few `workbox` packages and that is it.

**Comment by alexander-akait** [comment]
@srmagura Another idea - you have a component/module which import `react-dom` with another version, if you provide reproducible test repo I can investigate it and say that is wrong here (you can send me it on email, if you want I can sing CLA/NDA)

Can you modify:
```
vendor: {
  layer: layer => !layer || layer === 'app',
  test: /[\\/]node_modules[\\/]/,
  name: 'vendor',
  chunks: 'initial',
},
```

to

```
vendor: {
  layer: layer => !layer || layer === 'app',
  test: /[\\/]node_modules[\\/]/,
  name: 'vendor',
  chunks: 'initial',
  reuseExistingChunk: true,
  priority: -10
},
```

**Comment by srmagura** [comment]
I changed the code to:

```js
vendor: {
  layer: layer => !layer || layer === 'app',
  test: /[\\/]node_modules[\\/]/,
  name: 'vendor',

  chunks: 'all',
  reuseExistingChunk: true,
  priority: -10,
},
```

and now react-dom is only included in the big vendor chunk 🙂.

Then I tried changing `chunks` from `all` back to `initial`. Now `react-dom` is included in two chunks, which will cause issues at runtime.

We are going to continue compiling the service worker via separate webpack config since this seems less likely to cause issues. That said, I think I could create a minimal repro of this if that would be helpful.

(Note to self: `6acaf7a75`)
